Support alternative configurations for web content restrictions

davidquesada opened the pull request:
#### 2fe72762174e1f9efb5e71455bf41bcb82c94c83
<pre>
Support alternative configurations for web content restrictions
<a href="https://bugs.webkit.org/show_bug.cgi?id=291915">https://bugs.webkit.org/show_bug.cgi?id=291915</a>
<a href="https://rdar.apple.com/149723541">rdar://149723541</a>

Reviewed by Sihui Liu.

Make it possible for an app to direct WebContentRestrictions to a particular path
to load content restriction configurations from. This functionality is exposed on
_WKWebsiteDataStoreConfiguration, since the restriction configuration is applied to
all networking within that session. This reverses some of the singleton-ization
introduced in <a href="https://commits.webkit.org/292963@main">https://commits.webkit.org/292963@main</a>, since the ability to configure a
particular policy for web content restrictions is exposed by creating a new instance of
WCRBrowserEngineClient with a path. It&apos;s still beneficial to reuse instances of these
classes, but now there may need to be more than one in the case where one process uses
multiple configurations for different browsing contexts.

* Source/WebCore/PAL/pal/spi/cocoa/WebContentRestrictionsSPI.h:
* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::create):
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::webContentRestrictionsConfigurationPath const):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
(WebCore::ContentFilterUnblockHandler::configurationPath const):
(WebCore::ContentFilterUnblockHandler::setConfigurationPath):
* Source/WebCore/platform/PlatformContentFilter.h:
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync const):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::enabled const):
(WebCore::ParentalControlsContentFilter::ParentalControlsContentFilter):
(WebCore::ParentalControlsContentFilter::responseReceived):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::wcrBrowserEngineClientEnabled):
(WebCore::ParentalControlsURLFilter::filterWithConfigurationPath):
(WebCore::ParentalControlsURLFilter::ParentalControlsURLFilter):
(WebCore::ParentalControlsURLFilter::singleton):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::webContentRestrictionsConfigurationPath const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_dataStoreIdentifier):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::webContentRestrictionsConfigurationFile const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKErrorPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration webContentRestrictionsConfigurationURL]):
(-[_WKWebsiteDataStoreConfiguration setWebContentRestrictionsConfigurationURL:]):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::didHandleContentFilterUnblockNavigation):
    When handling a request from the web process to unblock a website, read the
    configuration path from this web view&apos;s data store, then hand it to the
    ContentFilterUnblockHandler. It didn&apos;t seem wise to make this property
    serializable / deserializable, since that could give a compromised web
    process the ability to direct the UI process to perform IO on an
    attacker-controlled path.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::setWebContentRestrictionsConfigurationFile):
(WebKit::WebsiteDataStoreConfiguration::webContentRestrictionsConfigurationFile const):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebContentRestrictions.mm: Added.
(-[WebContentRestrictionsNavigationDelegate webView:didFailProvisionalNavigation:withError:]):
(permissiveConfiguration):
(restrictiveConfiguration):
(TEST(WebContentRestrictions, ConfigurationFileDoesNotExist)):
(TEST(WebContentRestrictions, ConfigurationFileIsPermissive)):
(TEST(WebContentRestrictions, ConfigurationFileIsRestrictive)):
(TEST(WebContentRestrictions, MultipleConfigurationFiles)):

Canonical link: <a href="https://commits.webkit.org/294381@main">https://commits.webkit.org/294381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2bbdc879c9ae55a72332cb7072246f98875b13e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21274 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106764 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52240 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21582 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29795 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77381 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34413 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104613 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16668 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57718 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9777 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51587 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86361 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109117 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86355 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29100 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85918 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21869 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30664 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8378 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22891 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28667 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33956 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28478 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31801 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

davidquesada commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -27,24 +27,27 @@
     void isURLAllowed(const URL&, CompletionHandler<void(bool, NSData *)>&&);
     void allowURL(const URL&, CompletionHandler<void(bool)>&&);
 
+    ParentalControlsURLFilter(const String& configurationPath);
```
I know I shouldn't _need_ to make this initializer public, but calling `makeUniqueRef<ParentalControlsURLFilter>(key)` still complained about the use of a private constructor. I'm sure there's some combination of #includes I'm missing or some change I need to make to the `friend` declaration to make that work, but I couldn't figure it out.

whsieh replied:
I think akin to this should work?

```
friend UniqueRef<NetworkExtensionContentFilter> WTF::makeUniqueRefWithoutFastMallocCheck<NetworkExtensionContentFilter>();
```

...that said, making it public is a viable option too.

davidquesada replied:
Yeah, I was trying to look at the NetworkExtensionContentFilter as a reference, and I think this looks the same, with the exception of the argument added to the constructor:
```
friend UniqueRef<ParentalControlsURLFilter> WTF::makeUniqueRefWithoutFastMallocCheck<ParentalControlsURLFilter>(const String&);
```

I also tried slapping some `WTF::`s on all the types, but that didn't seem to help. (I also tried some other things to make this class look more like NetworkExtensionContentFilter -- marking it `final`, making the class TZone  allocated instead of fast allocated, adding all the WTF imports that NetworkExtensionContentFilter.h has, but none of those were the missing ingredient.)

szewai replied:
If we don't expect `ParentalControlsURLFilter` to be created outside of this class, maybe we should just do `UniqueRef<ParentalControlsURLFilter>(*new ParentalControlsURLFilter(path));` in ParentalControlsURLFilter.mm instead of making constructor public (and remove the `friend UniqueRef`).

szewai commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -27,24 +27,27 @@
 
 #if HAVE(WEBCONTENTRESTRICTIONS)
 
+
```
Additional blank line

szewai commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -27,24 +27,27 @@
-    ParentalControlsURLFilter();
 
     RetainPtr<WCRBrowserEngineClient> m_wcrBrowserEngineClient;
+    String m_configurationPath;
```
This is not used?

szewai commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -179,7 +179,7 @@
 #if HAVE(WEBCONTENTRESTRICTIONS)
     if (m_evaluatedURL) {
-        WebCore::ParentalControlsURLFilter::singleton().allowURL(*m_evaluatedURL, [decisionHandler = WTFMove(decisionHandler)](bool didAllow) {
+        WebCore::ParentalControlsURLFilter::filterWithConfigurationPath({ }).allowURL(*m_evaluatedURL, [decisionHandler = WTFMove(decisionHandler)](bool didAllow) {
```
This function can be invoked in UI process (`WebFrameProxy::didHandleContentFilterUnblockNavigation`), so I think you need to pass the path here, or this will update the default file and the change will not be observed by network process.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -354,6 +359,9 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
 #endif
     Vector<size_t> m_memoryFootprintNotificationThresholds;
     std::optional<bool> m_defaultTrackingPreventionEnabledOverride;
+#if HAVE(WEBCONTENTRESTRICTIONS)
```
This should be probably guarded by `WEBCONTENTRESTRICTIONS_PATH_SPI`? Same for all other places

szewai commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2198,6 +2198,14 @@ bool NetworkResourceLoader::usesWebContentRestrictions()
+
+String NetworkResourceLoader::webContentRestrictionsConfigurationPath() const
+{
+    if (auto* session = protectedConnectionToWebProcess()->protectedNetworkProcess()->networkSession(sessionID()))
```
`CheckedPtr`

szewai commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -37,21 +37,42 @@
-    if (wcrBrowserEngineClientEnabled())
+    if (wcrBrowserEngineClientEnabled(configurationPath)) {
+#if HAVE(WEBCONTENTRESTRICTIONS_PATH_SPI)
+        m_wcrBrowserEngineClient = adoptNS([PAL::allocWCRBrowserEngineClientInstance() initWithConfigurationAtPath:configurationPath.isEmpty() ? nullptr : configurationPath.createNSString().get()]);
```
Shouldn't we only invoke `initWithConfigurationAtPath` if path is not empty?

davidquesada committed:
Support alternative configurations for web content restrictions
https://bugs.webkit.org/show_bug.cgi?id=291915
rdar://149723541

Reviewed by Sihui Liu.

Make it possible for an app to direct WebContentRestrictions to a particular path
to load content restriction configurations from. This functionality is exposed on
_WKWebsiteDataStoreConfiguration, since the restriction configuration is applied to
all networking within that session. This reverses some of the singleton-ization
introduced in https://commits.webkit.org/292963@main, since the ability to configure a
particular policy for web content restrictions is exposed by creating a new instance of
WCRBrowserEngineClient with a path. It's still beneficial to reuse instances of these
classes, but now there may need to be more than one in the case where one process uses
multiple configurations for different browsing contexts.

* Source/WebCore/PAL/pal/spi/cocoa/WebContentRestrictionsSPI.h:
* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::create):
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::webContentRestrictionsConfigurationPath const):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
(WebCore::ContentFilterUnblockHandler::configurationPath const):
(WebCore::ContentFilterUnblockHandler::setConfigurationPath):
* Source/WebCore/platform/PlatformContentFilter.h:
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync const):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::enabled const):
(WebCore::ParentalControlsContentFilter::ParentalControlsContentFilter):
(WebCore::ParentalControlsContentFilter::responseReceived):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::wcrBrowserEngineClientEnabled):
(WebCore::ParentalControlsURLFilter::filterWithConfigurationPath):
(WebCore::ParentalControlsURLFilter::ParentalControlsURLFilter):
(WebCore::ParentalControlsURLFilter::singleton):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::webContentRestrictionsConfigurationPath const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_dataStoreIdentifier):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::webContentRestrictionsConfigurationFile const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKErrorPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration webContentRestrictionsConfigurationURL]):
(-[_WKWebsiteDataStoreConfiguration setWebContentRestrictionsConfigurationURL:]):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::didHandleContentFilterUnblockNavigation):
    When handling a request from the web process to unblock a website, read the
    configuration path from this web view's data store, then hand it to the
    ContentFilterUnblockHandler. It didn't seem wise to make this property
    serializable / deserializable, since that could give a compromised web
    process the ability to direct the UI process to perform IO on an
    attacker-controlled path.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::setWebContentRestrictionsConfigurationFile):
(WebKit::WebsiteDataStoreConfiguration::webContentRestrictionsConfigurationFile const):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebContentRestrictions.mm: Added.
(-[WebContentRestrictionsNavigationDelegate webView:didFailProvisionalNavigation:withError:]):
(permissiveConfiguration):
(restrictiveConfiguration):
(TEST(WebContentRestrictions, ConfigurationFileDoesNotExist)):
(TEST(WebContentRestrictions, ConfigurationFileIsPermissive)):
(TEST(WebContentRestrictions, ConfigurationFileIsRestrictive)):
(TEST(WebContentRestrictions, MultipleConfigurationFiles)):

Canonical link: https://commits.webkit.org/294381@main
